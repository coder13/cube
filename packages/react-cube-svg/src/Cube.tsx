import { Cube as CubeState, Face as FaceType } from "@cp/lib";
import Face from "./Face";

/* Colors: */
const Black = "#000000";
const White = "#FFFFFF";
const Yellow = "#FEFE00";
const Red = "#EE0000";
const Orange = "#FFA100";
const Blue = "#0000F2";
const Green = "#00D800";

export type ColorScheme = Record<FaceType, string>;

export interface CubeProps {
  scramble?: string;
  puzzle?: number;
  size?: number;
  lineWidth?: number;
  colorScheme?: ColorScheme;
  lineColor?: string;
  spacing?: number;
}

const Faces: Array<{
  face: FaceType;
  x: number;
  y: number;
}> = [
  {
    face: "up",
    x: 3,
    y: 0,
  },
  {
    face: "left",
    x: 0,
    y: 3,
  },
  {
    face: "front",
    x: 3,
    y: 3,
  },
  {
    face: "down",
    x: 3,
    y: 6,
  },
  {
    face: "right",
    x: 6,
    y: 3,
  },
  {
    face: "back",
    x: 9,
    y: 3,
  },
];

export const CubeRenderer = ({
  scramble = "",
  puzzle = 3,
  size = 200,
  colorScheme = {
    up: White,
    down: Yellow,
    front: Green,
    back: Blue,
    left: Orange,
    right: Red,
  },
  lineColor = Black,
  lineWidth = 1,
  spacing = 0.9,
  ...props
}: CubeProps) => {
  const state = new CubeState(scramble);

  const getFace = (f: FaceType) =>
    state.getFace(f).map((row) => row.map((s) => colorScheme[s]));

  return (
    <svg
      width={size}
      height={(size * 3) / 4}
      viewBox={`0 0 ${puzzle * 4} ${puzzle * 3}`}
      {...props}
    >
      {Faces.map(({ face, x, y }) => (
        <Face
          key={face}
          x={x}
          y={y}
          puzzle={puzzle}
          size={spacing}
          state={getFace(face)}
          lineColor={lineColor}
          lineWidth={lineWidth}
        />
      ))}
    </svg>
  );
};
