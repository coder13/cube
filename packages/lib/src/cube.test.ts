import Cube from "./cube";

import { describe, expect, test } from "bun:test";

describe("Cube", () => {
  test("Can do basic moves", () => {
    const cube = new Cube("U");
    expect(cube).toMatchSnapshot();
    cube.doMoves("R");
    expect(cube).toMatchSnapshot();
    cube.doMoves("F");
    expect(cube).toMatchSnapshot();
    cube.doMoves("D");
    expect(cube).toMatchSnapshot();
    cube.doMoves("L");
    expect(cube).toMatchSnapshot();
    cube.doMoves("B");
    expect(cube).toMatchSnapshot();
  });

  test("Cube.getFace", () => {
    const cube = new Cube();
    expect(cube.getFace("up")).toMatchSnapshot();
    expect(cube.getFace("right")).toMatchSnapshot();
    expect(cube.getFace("front")).toMatchSnapshot();
  });

  test("U", () => {
    const cube = new Cube("U");
    expect(cube.getFace("up")).toMatchSnapshot();
    expect(cube.getFace("down")).toMatchSnapshot();
    expect(cube.getFace("left")).toMatchSnapshot();
    expect(cube.getFace("right")).toMatchSnapshot();
    expect(cube.getFace("front")).toMatchSnapshot();
    expect(cube.getFace("back")).toMatchSnapshot();
  });
  test("R", () => {
    const cube = new Cube("R");
    expect(cube.getFace("up")).toMatchSnapshot();
    expect(cube.getFace("down")).toMatchSnapshot();
    expect(cube.getFace("left")).toMatchSnapshot();
    expect(cube.getFace("right")).toMatchSnapshot();
    expect(cube.getFace("front")).toMatchSnapshot();
    expect(cube.getFace("back")).toMatchSnapshot();
  });

  test("F", () => {
    const cube = new Cube("F");
    expect(cube.getFace("up")).toMatchSnapshot();
    expect(cube.getFace("down")).toMatchSnapshot();
    expect(cube.getFace("left")).toMatchSnapshot();
    expect(cube.getFace("right")).toMatchSnapshot();
    expect(cube.getFace("front")).toMatchSnapshot();
    expect(cube.getFace("back")).toMatchSnapshot();
  });

  test("D", () => {
    const cube = new Cube("D");
    expect(cube.getFace("up")).toMatchSnapshot();
    expect(cube.getFace("down")).toMatchSnapshot();
    expect(cube.getFace("left")).toMatchSnapshot();
    expect(cube.getFace("right")).toMatchSnapshot();
    expect(cube.getFace("front")).toMatchSnapshot();
    expect(cube.getFace("back")).toMatchSnapshot();
  });
  test("L", () => {
    const cube = new Cube("L");
    expect(cube.getFace("up")).toMatchSnapshot();
    expect(cube.getFace("down")).toMatchSnapshot();
    expect(cube.getFace("left")).toMatchSnapshot();
    expect(cube.getFace("right")).toMatchSnapshot();
    expect(cube.getFace("front")).toMatchSnapshot();
    expect(cube.getFace("back")).toMatchSnapshot();
  });

  test("B", () => {
    const cube = new Cube("B");
    expect(cube.getFace("up")).toMatchSnapshot();
    expect(cube.getFace("down")).toMatchSnapshot();
    expect(cube.getFace("left")).toMatchSnapshot();
    expect(cube.getFace("right")).toMatchSnapshot();
    expect(cube.getFace("front")).toMatchSnapshot();
    expect(cube.getFace("back")).toMatchSnapshot();
  });

  test("scramble", () => {
    const cube = new Cube(
      "D2 F2 L B2 U2 R' B2 L' R U2 B U' R' F' L F' D U2 F'"
    );
    expect(cube).toMatchSnapshot();
    expect(cube.getFace("up")).toMatchSnapshot();
    expect(cube.getFace("left")).toMatchSnapshot();
    expect(cube.getFace("front")).toMatchSnapshot();
    expect(cube.getFace("right")).toMatchSnapshot();
    expect(cube.getFace("back")).toMatchSnapshot();
    expect(cube.getFace("down")).toMatchSnapshot();
  });
});
