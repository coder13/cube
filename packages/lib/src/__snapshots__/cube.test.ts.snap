// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Cube Can do basic moves 1`] = `
Cube {
  "centers": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "corners": {
    "orient": [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "perm": [
      3,
      0,
      1,
      2,
      4,
      5,
      6,
      7,
    ],
  },
  "edges": {
    "orient": [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "perm": [
      3,
      0,
      1,
      2,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
    ],
  },
  "moves": " U",
}
`;

exports[`Cube Can do basic moves 2`] = `
Cube {
  "centers": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "corners": {
    "orient": [
      0,
      2,
      1,
      0,
      0,
      1,
      2,
      0,
    ],
    "perm": [
      3,
      1,
      6,
      2,
      4,
      0,
      5,
      7,
    ],
  },
  "edges": {
    "orient": [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "perm": [
      3,
      6,
      1,
      2,
      4,
      0,
      9,
      7,
      8,
      5,
      10,
      11,
    ],
  },
  "moves": " U R",
}
`;

exports[`Cube Can do basic moves 3`] = `
Cube {
  "centers": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "corners": {
    "orient": [
      0,
      2,
      2,
      1,
      0,
      1,
      2,
      1,
    ],
    "perm": [
      3,
      1,
      2,
      7,
      4,
      0,
      6,
      5,
    ],
  },
  "edges": {
    "orient": [
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      1,
      0,
      0,
      1,
      0,
    ],
    "perm": [
      3,
      6,
      7,
      2,
      4,
      0,
      1,
      10,
      8,
      5,
      9,
      11,
    ],
  },
  "moves": " U R F",
}
`;

exports[`Cube Can do basic moves 4`] = `
Cube {
  "centers": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "corners": {
    "orient": [
      0,
      2,
      2,
      1,
      1,
      2,
      1,
      0,
    ],
    "perm": [
      3,
      1,
      2,
      7,
      0,
      6,
      5,
      4,
    ],
  },
  "edges": {
    "orient": [
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      1,
      0,
      1,
      0,
      0,
    ],
    "perm": [
      3,
      6,
      7,
      2,
      4,
      0,
      1,
      10,
      5,
      9,
      11,
      8,
    ],
  },
  "moves": " U R F D",
}
`;

exports[`Cube Can do basic moves 5`] = `
Cube {
  "centers": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "corners": {
    "orient": [
      2,
      2,
      2,
      2,
      2,
      2,
      1,
      2,
    ],
    "perm": [
      0,
      1,
      2,
      3,
      4,
      6,
      5,
      7,
    ],
  },
  "edges": {
    "orient": [
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      1,
    ],
    "perm": [
      3,
      6,
      7,
      4,
      8,
      0,
      1,
      2,
      5,
      9,
      11,
      10,
    ],
  },
  "moves": " U R F D L",
}
`;

exports[`Cube Can do basic moves 6`] = `
Cube {
  "centers": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "corners": {
    "orient": [
      1,
      0,
      2,
      2,
      0,
      1,
      1,
      2,
    ],
    "perm": [
      1,
      6,
      2,
      3,
      0,
      4,
      5,
      7,
    ],
  },
  "edges": {
    "orient": [
      1,
      0,
      1,
      0,
      1,
      1,
      1,
      0,
      1,
      1,
      0,
      1,
    ],
    "perm": [
      0,
      6,
      7,
      4,
      3,
      5,
      1,
      2,
      8,
      9,
      11,
      10,
    ],
  },
  "moves": " U R F D L B",
}
`;

exports[`Cube Cube.getFace 1`] = `
[
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "up",
  ],
]
`;

exports[`Cube Cube.getFace 2`] = `
[
  [
    "right",
    "right",
    "right",
  ],
  [
    "right",
    "right",
    "right",
  ],
  [
    "right",
    "right",
    "right",
  ],
]
`;

exports[`Cube Cube.getFace 3`] = `
[
  [
    "front",
    "front",
    "front",
  ],
  [
    "front",
    "front",
    "front",
  ],
  [
    "front",
    "front",
    "front",
  ],
]
`;

exports[`Cube U 1`] = `
[
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "up",
  ],
]
`;

exports[`Cube U 2`] = `
[
  [
    "down",
    "down",
    "down",
  ],
  [
    "down",
    "down",
    "down",
  ],
  [
    "down",
    "down",
    "down",
  ],
]
`;

exports[`Cube U 3`] = `
[
  [
    "front",
    "front",
    "front",
  ],
  [
    "left",
    "left",
    "left",
  ],
  [
    "left",
    "left",
    "left",
  ],
]
`;

exports[`Cube U 4`] = `
[
  [
    "back",
    "back",
    "back",
  ],
  [
    "right",
    "right",
    "right",
  ],
  [
    "right",
    "right",
    "right",
  ],
]
`;

exports[`Cube U 5`] = `
[
  [
    "right",
    "right",
    "right",
  ],
  [
    "front",
    "front",
    "front",
  ],
  [
    "front",
    "front",
    "front",
  ],
]
`;

exports[`Cube U 6`] = `
[
  [
    "left",
    "left",
    "left",
  ],
  [
    "back",
    "back",
    "back",
  ],
  [
    "back",
    "back",
    "back",
  ],
]
`;

exports[`Cube R 1`] = `
[
  [
    "up",
    "up",
    "front",
  ],
  [
    "up",
    "up",
    "front",
  ],
  [
    "up",
    "up",
    "front",
  ],
]
`;

exports[`Cube R 2`] = `
[
  [
    "down",
    "down",
    "back",
  ],
  [
    "down",
    "down",
    "back",
  ],
  [
    "down",
    "down",
    "back",
  ],
]
`;

exports[`Cube R 3`] = `
[
  [
    "left",
    "left",
    "left",
  ],
  [
    "left",
    "left",
    "left",
  ],
  [
    "left",
    "left",
    "left",
  ],
]
`;

exports[`Cube R 4`] = `
[
  [
    "right",
    "right",
    "right",
  ],
  [
    "right",
    "right",
    "right",
  ],
  [
    "right",
    "right",
    "right",
  ],
]
`;

exports[`Cube R 5`] = `
[
  [
    "front",
    "front",
    "down",
  ],
  [
    "front",
    "front",
    "down",
  ],
  [
    "front",
    "front",
    "down",
  ],
]
`;

exports[`Cube R 6`] = `
[
  [
    "up",
    "back",
    "back",
  ],
  [
    "up",
    "back",
    "back",
  ],
  [
    "up",
    "back",
    "back",
  ],
]
`;

exports[`Cube F 1`] = `
[
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "up",
  ],
  [
    "left",
    "left",
    "left",
  ],
]
`;

exports[`Cube F 2`] = `
[
  [
    "right",
    "right",
    "right",
  ],
  [
    "down",
    "down",
    "down",
  ],
  [
    "down",
    "down",
    "down",
  ],
]
`;

exports[`Cube F 3`] = `
[
  [
    "left",
    "left",
    "down",
  ],
  [
    "left",
    "left",
    "down",
  ],
  [
    "left",
    "left",
    "down",
  ],
]
`;

exports[`Cube F 4`] = `
[
  [
    "up",
    "right",
    "right",
  ],
  [
    "up",
    "right",
    "right",
  ],
  [
    "up",
    "right",
    "right",
  ],
]
`;

exports[`Cube F 5`] = `
[
  [
    "front",
    "front",
    "front",
  ],
  [
    "front",
    "front",
    "front",
  ],
  [
    "front",
    "front",
    "front",
  ],
]
`;

exports[`Cube F 6`] = `
[
  [
    "back",
    "back",
    "back",
  ],
  [
    "back",
    "back",
    "back",
  ],
  [
    "back",
    "back",
    "back",
  ],
]
`;

exports[`Cube D 1`] = `
[
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "up",
  ],
]
`;

exports[`Cube D 2`] = `
[
  [
    "down",
    "down",
    "down",
  ],
  [
    "down",
    "down",
    "down",
  ],
  [
    "down",
    "down",
    "down",
  ],
]
`;

exports[`Cube D 3`] = `
[
  [
    "left",
    "left",
    "left",
  ],
  [
    "left",
    "left",
    "left",
  ],
  [
    "back",
    "back",
    "back",
  ],
]
`;

exports[`Cube D 4`] = `
[
  [
    "right",
    "right",
    "right",
  ],
  [
    "right",
    "right",
    "right",
  ],
  [
    "front",
    "front",
    "front",
  ],
]
`;

exports[`Cube D 5`] = `
[
  [
    "front",
    "front",
    "front",
  ],
  [
    "front",
    "front",
    "front",
  ],
  [
    "left",
    "left",
    "left",
  ],
]
`;

exports[`Cube D 6`] = `
[
  [
    "back",
    "back",
    "back",
  ],
  [
    "back",
    "back",
    "back",
  ],
  [
    "right",
    "right",
    "right",
  ],
]
`;

exports[`Cube L 1`] = `
[
  [
    "back",
    "up",
    "up",
  ],
  [
    "back",
    "up",
    "up",
  ],
  [
    "back",
    "up",
    "up",
  ],
]
`;

exports[`Cube L 2`] = `
[
  [
    "front",
    "down",
    "down",
  ],
  [
    "front",
    "down",
    "down",
  ],
  [
    "front",
    "down",
    "down",
  ],
]
`;

exports[`Cube L 3`] = `
[
  [
    "left",
    "left",
    "left",
  ],
  [
    "left",
    "left",
    "left",
  ],
  [
    "left",
    "left",
    "left",
  ],
]
`;

exports[`Cube L 4`] = `
[
  [
    "right",
    "right",
    "right",
  ],
  [
    "right",
    "right",
    "right",
  ],
  [
    "right",
    "right",
    "right",
  ],
]
`;

exports[`Cube L 5`] = `
[
  [
    "up",
    "front",
    "front",
  ],
  [
    "up",
    "front",
    "front",
  ],
  [
    "up",
    "front",
    "front",
  ],
]
`;

exports[`Cube L 6`] = `
[
  [
    "back",
    "back",
    "down",
  ],
  [
    "back",
    "back",
    "down",
  ],
  [
    "back",
    "back",
    "down",
  ],
]
`;

exports[`Cube B 1`] = `
[
  [
    "right",
    "right",
    "right",
  ],
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "up",
  ],
]
`;

exports[`Cube B 2`] = `
[
  [
    "down",
    "down",
    "down",
  ],
  [
    "down",
    "down",
    "down",
  ],
  [
    "left",
    "left",
    "left",
  ],
]
`;

exports[`Cube B 3`] = `
[
  [
    "up",
    "left",
    "left",
  ],
  [
    "up",
    "left",
    "left",
  ],
  [
    "up",
    "left",
    "left",
  ],
]
`;

exports[`Cube B 4`] = `
[
  [
    "right",
    "right",
    "down",
  ],
  [
    "right",
    "right",
    "down",
  ],
  [
    "right",
    "right",
    "down",
  ],
]
`;

exports[`Cube B 5`] = `
[
  [
    "front",
    "front",
    "front",
  ],
  [
    "front",
    "front",
    "front",
  ],
  [
    "front",
    "front",
    "front",
  ],
]
`;

exports[`Cube B 6`] = `
[
  [
    "back",
    "back",
    "back",
  ],
  [
    "back",
    "back",
    "back",
  ],
  [
    "back",
    "back",
    "back",
  ],
]
`;

exports[`Cube scramble 1`] = `
Cube {
  "centers": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "corners": {
    "orient": [
      1,
      1,
      2,
      0,
      0,
      1,
      1,
      0,
    ],
    "perm": [
      6,
      0,
      1,
      7,
      5,
      2,
      4,
      3,
    ],
  },
  "edges": {
    "orient": [
      0,
      1,
      1,
      1,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      1,
    ],
    "perm": [
      2,
      10,
      6,
      8,
      11,
      5,
      4,
      3,
      7,
      9,
      0,
      1,
    ],
  },
  "moves": " D2 F2 L B2 U2 R' B2 L' R U2 B U' R' F' L F' D U2 F'",
}
`;

exports[`Cube scramble 2`] = `
[
  [
    "front",
    "up",
    "left",
  ],
  [
    "back",
    "up",
    "front",
  ],
  [
    "down",
    "right",
    "right",
  ],
]
`;

exports[`Cube scramble 3`] = `
[
  [
    "right",
    "down",
    "front",
  ],
  [
    "left",
    "left",
    "up",
  ],
  [
    "back",
    "up",
    "front",
  ],
]
`;

exports[`Cube scramble 4`] = `
[
  [
    "left",
    "front",
    "back",
  ],
  [
    "left",
    "front",
    "back",
  ],
  [
    "left",
    "up",
    "left",
  ],
]
`;

exports[`Cube scramble 5`] = `
[
  [
    "up",
    "down",
    "up",
  ],
  [
    "left",
    "right",
    "right",
  ],
  [
    "down",
    "right",
    "front",
  ],
]
`;

exports[`Cube scramble 6`] = `
[
  [
    "back",
    "front",
    "down",
  ],
  [
    "back",
    "back",
    "down",
  ],
  [
    "up",
    "left",
    "right",
  ],
]
`;

exports[`Cube scramble 7`] = `
[
  [
    "up",
    "back",
    "back",
  ],
  [
    "right",
    "down",
    "down",
  ],
  [
    "down",
    "front",
    "right",
  ],
]
`;
